streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -648,55 +648,62 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
-                   case Left(outDone) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *> outgoing.offer(ZIO.succeed(Left(outDone)))
-                   case Right(outElem) =>
-                     for {
-                       p     <- Promise.make[Unit, OutElem2]
-                       latch <- Promise.make[Nothing, Unit]
-                       _     <- outgoing.offer(p.await.map(Right(_)))
-                       _ <- permits.withPermit {
-                              latch.succeed(()) *>
-                                ZIO.uninterruptibleMask { restore =>
-                                  restore(errorSignal.await).raceFirstAwait(
-                                    restore(f(outElem))
-                                      .catchAllCause(cause => failure.update(_ && cause) *> Exit.failCause(Cause.unit))
-                                  )
-                                }.foldCauseZIO(
-                                  _ => p.refailCause(Cause.unit) *> errorSignal.refailCause(Cause.unit),
-                                  p.succeed
-                                )
-                            }.forkIn(scope)
-                       _ <- latch.await
-                     } yield ()
-                 }
-               )
-               .forever
-               .interruptible
-               .forkIn(scope)
+        pull        <- (queueReader >>> self).toPullInAlt(scope)
+        _ <-
+          pull
+            .flatMap(outElem =>
+              for {
+                latch <- Promise.make[Nothing, Unit]
+                f <- permits
+                       .withPermit(
+                         latch.succeed(()) *> f(outElem)
+                           .catchAllCause(cause =>
+                             failure.update(_ && cause).unless(cause.isInterrupted) *>
+                               errorSignal.succeed(()) *>
+                               ZChannel.failLeftUnit
+                           )
+                       )
+                       .fork
+                _ <- latch.await
+                _ <- outgoing.offer(f)
+              } yield ()
+            )
+            .forever
+            .catchAllCause(cause =>
+              cause.failureOrCause match {
+                case Left(x: Left[OutErr, OutDone]) =>
+                  failure.update(_ && Cause.fail(x.value)) *>
+                    outgoing.offer(Fiber.done(ZChannel.failLeftUnit)) *>
+                    ZChannel.failUnit
+                case Left(x: Right[OutErr, OutDone]) =>
+                  permits.withPermits(n.toLong)(ZIO.unit) *>
+                    outgoing.offer(Fiber.fail(x.asInstanceOf[Either[Unit, OutDone]]))
+                case Right(cause) =>
+                  failure.update(_ && cause).unless(cause.isInterrupted) *>
+                    outgoing.offer(Fiber.done(ZChannel.failLeftUnit)) *>
+                    ZChannel.failUnit
+              }
+            )
+            .race(errorSignal.await)
```

Comment:
See my other comment, we should change this to `errorSignal.await.interruptible`